### DISCUSSION TITLE
How to avoid false positive typecheck errors when running golangci-lint

### DISCUSSION THREAD

**Discussion by nuhasha** [golangci/golangci-lint#5008] [Q&A] [answered]
I have a question: how do I run the `golangci-lint`? based on what mentioned in [4306](https://github.com/golangci/golangci-lint/pull/4306/files )

> Ensure you are not running an analysis on code that depends on files/packages outside the scope of the analyzed elements.

for example in my PR it gives an error on one of the variables in `main.go`

```Error: main.go:153:7: cfg.VariableX undefined (type Config has no field or method VariableX) (typecheck)```
  		
but VariableX is defined in `folder1/folder2/config.go`

the linting file is part of GitHub workflows `.github/workflows/linting.yaml`  so, I am not sure what I am doing wrong here.

Here is my repo structure

```
|main.go
|folder1
|    folder2
|        config.go
|.github
|     workflows
|          linting.yaml  
```

**Comment by ldez** [comment]
Hello,

first, `typecheck` has no false positive because it's not a linter, I recommend reading the following page to have more information about its behavior https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors

As explained in the doc:

> Of course, this is just as good as the compiler itself and a lot of compilation issues will not properly show where in the code your error lies.

So you have a problem inside your code and it's maybe not related to the log you provided.

I cannot help more without a concrete minimal reproducible example.

**Reply by nuhasha** [reply]
Hi @ldez  thank you for your response, the issue was fixed for me when I passed` --fast` flag.

**Reply by ldez** [reply]
This is not a fix, you just disabled 90% of the linters.

Please provide a minimal reproducible example.

**Reply by nuhasha** [reply]
`golangci-lint` is complaining about variables `x`, and `y` as  declared and not used: 
However, the struct Component is being used somewhere else in another file in the project. 

```go
package main

import "fmt"

type Component struct {
	Cloud1   int
	Cloud2 int
}

func Hello() {
	x := 5
	y := 6
	component := Component{
				Cloud1:   x,
				Cloud2: y,
	}
}

func main() {
	Hello()
}

```
The code is more complicated than that, as we assign the value for x and y through another function. and then we have couple of if/else to assign the values for Component based on some other outputs, but this is kind of get some idea of the issue I face.

**Reply by ldez** [reply]
The code you provided doesn't compile, can you give a code example that compiles and reproduces the problem?

**Reply by nuhasha** [reply]
Example: 
```
mkdir hello
cd hello
touch hello.go
go mod init example.com/hello
```
hello.go
```
package main

import "fmt"

type Component struct {
	Cloud1 int
	Cloud2 int
}

func Hello() {
	x := 5
	y := 6
	component := Component{
		Cloud1: x,
		Cloud2: y,
	}
	fmt.Println(component.Cloud1, component.Cloud2)
}

func main() {
	Hello()
}

```

**Reply by ldez** [reply]
I think you misunderstood my comment because this code compiles but there is no problem with `typecheck`.

I'm looking for something that reproduces your problem.

**Reply by nuhasha** [reply]
I get this error only when github actions run on PR, but when I run the `golangci-lint` locally in `vscode`, I don't get this error. 

My original code does not have compilation error as it builds successfully.

**Reply by ldez** [reply]
I need more information, without that it's possible to diagnose the problem.

So can you provide the content of the workflow file?

**Reply by ldez** [reply, marked as answer]
I saw the message you deleted, and now I know where your problem is.

You are using private modules but those modules are not available inside your CI, so you have `typechek` errors.

To validate that:
- you need to add a step to download your modules before running the linter (I think it will fail)
- After this new step add a step to build your program (I think it will fail too)

```yml
      - name: download dependencies
        run: go mod tidy
        
      - name: build
        run: go build .

      - name: golangci-lint
      # ...
```

**Reply by nuhasha** [reply]
Thank you @ldez  issues fixed, I no longer have the `typechek` error, and now I see same outputs when I run the workflow both locally or through GitHub actions on PR. 

from your previous comment regarding downloading dependencies:

```
  - name: download dependencies
        run: go mod tidy
```

I got an error: repo does not exist, so, I had to provide the token for another private repo, after which `golangci-lint` ran.

I appreciate your help in investigating issue.

**Reply by ldez** [reply]
I hope you enjoyed my work, please consider donating or asking your company to do so.
This will be appreciated, thank you :heart:

[![Open Collective](https://img.shields.io/badge/OpenCollective-Donate-blue?logo=opencollective&style=for-the-badge)](https://opencollective.com/golangci-lint)
[![GitHub Sponsors](https://img.shields.io/badge/GitHub-Donate-blue?logo=github&style=for-the-badge)](https://github.com/sponsors/golangci)
[![GitHub Sponsors](https://img.shields.io/badge/ldez-Donate-blue?logo=github&style=for-the-badge)](https://github.com/sponsors/ldez)
